### PR TITLE
Use `conda-forge-build-setup` for CI configuration

### PR DESCRIPTION
**Pull request by jakirkham**
This will switch us over to using [`conda-forge-build-setup`]( https://github.com/conda-forge/conda-forge-build-setup-feedstock ) to control CI configuration instead of doing it manually. In the long run, we plan to replace the bulk of the CI scripts both in staged-recipes and feedstocks with this package. It should make pinning workarounds for `conda-build` simpler to make, track, and remove without getting tangled up with other CI stuff. Also, this should help remove duplication between staged-recipes and conda-smithy templates.

cc @pelson @ocefpaf @msarahan @jjhelmus

### CODE REVIEW

**Comment by jakirkham** [comment]
@msarahan, would you mind double checking what we are doing here on Windows? It should just run this [script]( https://github.com/conda-forge/conda-forge-build-setup-feedstock/blob/1cd33c4db54ac6924d965c98c4ae62143cbafe6d/recipe/activate.bat ), but it doesn't seem to be doing that.

This is basically a simplified version of what @jjhelmus is doing in PR ( https://github.com/conda-forge/python-feedstock/pull/18 ). So, hopefully, this will be easier to debug.

**Comment by jakirkham** [comment]
Interestingly, it doesn't appear to be pinning `conda-build` correctly on Windows either.

**Comment by jjhelmus** [comment]
@jakirkham You could try running `conda config --get` before and after the activate on windows to dump configuration.  I'm working on getting a Windows VM up locally but it might take a bit.

**Comment by jakirkham** [comment]
> You could try running `conda config --get` before and after the activate on windows to dump configuration.

Ok, giving that a try.

> I'm working on getting a Windows VM up locally but it might take a bit.

Sounds great. Let me know if there is anything I can do to help.

**Comment by jakirkham** [comment]
Yeah, the activation script is definitely not being run.

**Comment by jakirkham** [comment]
@janschulz, if you have any ideas on activation scripts and Windows, we could really use your help over here.

**Comment by jakirkham** [comment]
> Interestingly, it doesn't appear to be pinning `conda-build` correctly on Windows either.

Wow, I think this was a massive oversight on my part. This I can fix. As for the activation stuff, I'm still at a loss.

**Comment by jjhelmus** [comment]
I don't think the windows conda packages are properly packaging the activate.bat file.  Unpack the package and there is a Library/etc/activate.d directory with the activate.bat file and meta.yaml file concatenated. That seems a bit odd.

**Comment by jakirkham** [comment]
Yep, I think we messed up with these [lines]( https://github.com/conda-forge/conda-forge-build-setup-feedstock/blob/1cd33c4db54ac6924d965c98c4ae62143cbafe6d/recipe/meta.yaml#L8-L9 ). @janschulz suggested something similar in another thread. Though I'm starting to think this use case would be better off with post-link scripts anyways. ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/issues/3 )

**Comment by janschulz** [comment]
Just for the record: it seesm that with the currently released conda version, activation script which want to set a environment variable other than PATH wont get through as the activate.d action is don in the `setlocal` part of the script and the `endlocal` line only sets the following:

```
    SET "PATH=%PATH%"
    SET "PROMPT=%PROMPT%"
    SET "CONDA_DEFAULT_ENV=%CONDA_DEFAULT_ENV%"
    SET "CONDA_PATH_BACKUP=%CONDA_PATH_BACKUP%"
```

The fix is in https://github.com/conda/conda/commit/f531272c24b8c123cab9ddd7c66154cba277d5cc, which is commited on 15 april but seems not to be on the [4.0.x branch](https://github.com/conda/conda/commits/4.0.x) :-(

If you want to test it localy, you can change your current script to read: 

```
rem echo on
ENDLOCAL & (
    SET "PATH=%PATH%"
    SET "PROMPT=%PROMPT%"
    SET "CONDA_DEFAULT_ENV=%CONDA_DEFAULT_ENV%"
    SET "CONDA_PATH_BACKUP=%CONDA_PATH_BACKUP%"
    
)

REM ADD THIS...

REM Run any activate scripts
IF EXIST "%CONDA_DEFAULT_ENV%\etc\conda\activate.d" (
    PUSHD "%CONDA_DEFAULT_ENV%\etc\conda\activate.d"
    FOR %%g in (*.bat) DO @CALL "%%g"
    POPD
)

REM ...UNTIL HERE...

EXIT /B
```

**Comment by jjhelmus** [comment]
> Yep, I think we messed up with these lines.

Agreed, I think those lines should use `%PREFIX%` not `%LIBRARY_PREFIX%` but I don't know why everything in the recipe directory is getting concatenated together in the script created.

**Comment by janschulz** [comment]
> but I don't know why everything in  recipe directory is getting concatenated together in the script created.

I also had that.Not sure if that was a problem where the copy line had a wrong slash in one of the filenames.

These lines work: https://github.com/jakirkham/staged-recipes/pull/11/files#diff-a787ff5ab7de2a3624df9a71802c4cd8R39

**Comment by msarahan** [comment]
This is not in Conda 4.0.x because the activate scripts live in conda-env until conda 4.0.x.  I don't recall whether those changes ever made it into a conda-env release.

File: .CI/build_all [hunk 1]
```diff
@@ -11,7 +11,8 @@ if [ -z "$GH_TOKEN" ]; then
     conda config --set show_channel_urls true
     conda config --add channels conda-forge
     conda install --yes --quiet conda-build-all
-    conda install --yes --quiet conda-build=1.20.0 jinja2 anaconda-client setuptools
+    conda install --yes --quiet conda-forge-build-setup
+    source activate root
```
Comment (pelson) [inline .CI/build_all]: Let's skip the activate step, and just install a script from `conda-forge-build-setup` which we run. It wouldn't be such a problem if that script did stateful things like updating configs, or running other conda commands then.

Comment (jakirkham) [inline .CI/build_all]: SGTM

**Comment by jakirkham** [comment]
This is a desirable failure. 😆 Basically, this shows us that a package that requires `setuptools` and doesn't list it will fail trying to import it. The same story is true of `pip` or `wheel`. Will add back the dependency to make sure this will work as expected.

**Comment by jakirkham** [comment]
Should add that the intended downgrades are happening. Also, I checked for environment variables to make sure they were being set correctly.

**Comment by jakirkham** [comment]
This seems to be working correctly. Removing the `toolz` recipe that was being used for testing.

**Comment by jakirkham** [comment]
Please feel free to share any more relevant feedback about this change. If it seems acceptable, feel free to merge.

We plan to try this out at staged-recipes. This will give us a chance to work out any funny things in an active environment. Once we have ironed this change out, we will add this to `conda-smithy` and do a new release.

Going forward all pinnings for `conda`, `conda-build`, `conda-build-all`, etc. Will be managed via [`conda-forge-build-setup`]( https://github.com/conda-forge/conda-forge-build-setup-feedstock ). This will make it easier to track issues that arise at one location and deploy them to others easily. By doing so, it frees us from requiring a `conda-smithy` release for each change of this sort. Hopefully, it will cut down the effort spent fixing these sorts of issues.

File: .CI/build_all [hunk 2]
```diff
@@ -11,7 +11,8 @@ if [ -z "$GH_TOKEN" ]; then
     conda config --set show_channel_urls true
     conda config --add channels conda-forge
     conda install --yes --quiet conda-build-all
-    conda install --yes --quiet conda-build=1.20.0 jinja2 anaconda-client setuptools
+    conda install --yes --quiet conda-forge-build-setup
+    source "`conda info --root`/bin/run_conda_forge_build_setup"
```
Comment (pelson) [inline .CI/build_all]: Surely the `conda info --root` is superfluous - conda is already on the PATH, so it should just find it...?

File: appveyor.yml [hunk 3]
```diff
@@ -71,16 +71,12 @@ install:
           $env:path="$root;$root\Scripts;$root\Library\bin;$($env:path)"
     - cmd: conda config --set show_channel_urls true
     - cmd: conda update --yes --quiet conda
-    - cmd: set PYTHONUNBUFFERED=1
 
     - cmd: conda config --add channels conda-forge
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: obvci_install_conda_build_tools.py
```
Comment (pelson) [inline appveyor.yml]: This line can probably go (xref: https://github.com/pelson/Obvious-CI/blob/master/scripts/obvci_install_conda_build_tools.py)

File: scripts/run_docker_build.sh [hunk 4]
```diff
@@ -30,19 +30,15 @@ fi
...
 
 # A lock sometimes occurs with incomplete builds. The lock file is stored in build_artefacts.
 conda clean --lock
 
 conda update conda conda-build
-conda install anaconda-client conda-build-all
-conda info
+conda install conda-build-all
+conda install conda-forge-build-setup
+source "\`conda info --root\`/bin/run_conda_forge_build_setup"
```
Comment (pelson) [inline scripts/run_docker_build.sh]: Ditto, should already be on the PATH.

**Comment by pelson** [comment]
I agree with the direction. I have a few specific changes I would make, and will propose them shortly in a follow-on PR.

:+1: - awesome @jakirkham!

**Comment by jakirkham** [comment]
Thanks, @pelson. So, we need to use `source` to make sure the environment variables the scripts set do get exported to the shell running the scripts. It is for this reason I used the full path.

File: appveyor.yml (hunk 3 above)
Comment (jakirkham) [inline appveyor.yml]: I see. Sure. Let's get rid of it then.

**Comment by jakirkham** [comment]
See this PR ( https://github.com/conda-forge/staged-recipes/pull/743 ).

**Pull request merged**